tower-http/src/classify.rs

```diff
@@ -0,0 +1,170 @@
+//! Tools for classifying responses as either success or failure.
+
+use http::{HeaderMap, Response, StatusCode};
+use http_body::Body;
+use std::{convert::Infallible, marker::PhantomData};
+
+/// Trait for classifying responses as either success or failure.
+pub trait ClassifyResponse {
+    /// The output of doing the classification.
+    ///
+    /// This could be an HTTP status code or some other kind of status.
+    type Output;
+
+    /// The type used to classify the end of streaming (Eos) responses.
+    type ClassifyEos: ClassifyEos<Output = Self::Output>;
+
+    /// Perform the classification.
+    fn classify_response<B>(
+        &mut self,
+        res: &Response<B>,
+    ) -> ClassifiedNowOrLater<Self::Output, Self::ClassifyEos>
+    where
+        B: Body;
+}
+
+/// Trait for classifying end of streams ("Eos") as either success or failure.
+pub trait ClassifyEos {
+    /// The output of doing the classification.
+    ///
+    /// This could be an HTTP status code or some other kind of status.
+    type Output;
+
+    /// Perform the classification.
+    fn classify_eos(&mut self, trailers: Option<&HeaderMap>) -> Classification<Self::Output>;
+}
+
+/// Result of doing a classification.
+#[derive(Debug)]
+// really don't like this name... :(
+pub enum ClassifiedNowOrLater<T, ClassifyEos> {
+    /// The response was able to be classified immediately.
+    Ready(Classification<T>),
+    /// We have to wait until the end of a streaming response to classify it.
+    RequiresEos(ClassifyEos),
+}
+
+/// A classification of a response.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Classification<T> {
+    /// The response was a success.
+    Success(T),
+    /// The response was a failure.
+    Failure(T),
+}
```

Comment:
Hm yeah I think removing the success value is fine. The tracing and metrics layers are still able to access the status if necessary.

I also went ahead and built a retry policy based on your suggestion. Seems like it could work.